### PR TITLE
E2 1899 network layer request infra

### PR DESCRIPTION
**Pull request by hrishileanplum**
https://leanplum.atlassian.net/browse/E2-1899
WIP for network infrastructure.

### CODE REVIEW

File: Leanplum/Leanplum.h [hunk 1]
```diff
@@ -14,6 +14,33 @@ FOUNDATION_EXPORT double LeanplumVersionNumber;
 //! Project version string for Leanplum.
 FOUNDATION_EXPORT const unsigned char LeanplumVersionString[];
 
-// In this header, you should import all the public headers of your framework using statements like #import <Leanplum/PublicHeader.h>
 
+@interface Leanplum : NSObject
```
Comment (e7mac) [inline Leanplum/Leanplum.h]: If we make the Leanplum object have class functions, can we ensure that there is no state that we store in this class? If we do want to store state, we should do it in a different class and just call it from this one

File: Leanplum/Leanplum.m [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+                withSuccess:(void (^)(void))success
+                withFailure:(void (^)(NSError *error))failure];
+}
+
++ (void)setUserIdInternal:(NSString *)userId
+           withAttributes:(NSDictionary *)attributes
+              withSuccess:(void (^)(void))success
+              withFailure:(void (^)(NSError *error))failure
+{
+    // Some clients are calling this method with NSNumber. Handle it gracefully.
```
Comment (e7mac) [inline Leanplum/Leanplum.m]: nice!

File: Leanplum/Utils/LPUtils.h [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+//  Unless required by applicable law or agreed to in writing,
+//  software distributed under the License is distributed on an
+//  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+//  KIND, either express or implied.  See the License for the
+//  specific language governing permissions and limitations
+//  under the License.
+
+#import <Foundation/Foundation.h>
+
+@interface LPUtils : NSObject
```
Comment (e7mac) [inline Leanplum/Utils/LPUtils.h]: Could we think of a way to not have a generic `LPUtils` ?
Maybe split these into `LPObjectValidation` and others?

File: Leanplum/Utils/LPUtils.h [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
++ (NSString *)md5OfData:(NSData *)data;
+
+/**
+ * Returns base64 encoded string from NSData. Convenience method
+ * that supports iOS6.
+ */
++ (NSString *)base64EncodedStringFromData:(NSData *)data;
+
+/**
+ * Initialize exception handling
```
Comment (e7mac) [inline Leanplum/Utils/LPUtils.h]: These wrappers for ExceptionHandling should instead just use the LPExceptionHandler directly, doesnt

File: Leanplum/Leanplum.h (hunk 1 above)
Comment (hrishileanplum) [reply Leanplum/Leanplum.h]: sure. We will keep that in mind. Once we have caching layer started.

File: Leanplum/Utils/LPUtils.h (hunk 4 above)
Comment (hrishileanplum) [reply Leanplum/Utils/LPUtils.h]: Good point.

File: Leanplum/Utils/LPUtils.h (hunk 3 above)
Comment (hrishileanplum) [reply Leanplum/Utils/LPUtils.h]: I think, we would still need a generic LPUtils. Lets revisit this later.

File: Leanplum/Leanplum.h [hunk 5]
```diff
@@ -14,6 +14,33 @@ FOUNDATION_EXPORT double LeanplumVersionNumber;
...
+      withSuccess:(void (^)(void))success
+      withFailure:(void (^)(NSError *error))failure;
+
+/**
+ * Updates a user ID after session start with a dictionary of user attributes.
+ */
++ (void)setUserId:(NSString *)userId withUserAttributes:(NSDictionary *)attributes
+      withSuccess:(void (^)(void))success
+      withFailure:(void (^)(NSError *error))failure;
+
```
Comment (gg4race) [inline Leanplum/Leanplum.h]: I feel like the naming here is kind of strange. userId is a part of user attributes, why do we have to have a separate function just for setUserId? Can we have a setUserAttributes(userId) function with just user Id as param? We can have 3 setUserAttributes with varying params here then. But if we have to keep the same methods as the old SDK then this is fine

File: Leanplum/Leanplum.m [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        attributes = @{};
+    }
+    [LPUserApi setUsersAttributes:userId withUserAttributes:attributes success:^{
+        success();
+    } failure:^(NSError *error) {
+        failure(error);
+    }];
+}
+
++ (NSDictionary *)validateAttributes:(NSDictionary *)attributes named:(NSString *)argName
```
Comment (gg4race) [inline Leanplum/Leanplum.m]: can we put comment //copied from old SDK if we copy code exactly from old sdk? This way there are a lot fewer questions about certain things are done this way

File: Leanplum/Utils/LPExceptionHandler.h [hunk 7]
```diff
@@ -0,0 +1,35 @@
+//
+//  LPExceptionHandler.h
```
Comment (gg4race) [inline Leanplum/Utils/LPExceptionHandler.h]: minor thing, the headers look really different when copied over from old sdk, do we care about this?

File: Leanplum/Leanplum.m (hunk 6 above)
Comment (hrishileanplum) [reply Leanplum/Leanplum.m]: Thanks for the comment @gg4race . As per our meeting discussion. We will cherry pick and not reinvent wheel at some places like these . And lets review the code even thought its copied form somewhere else. No need to add comment as //copied etc.

File: Leanplum/Utils/LPExceptionHandler.h (hunk 7 above)
Comment (hrishileanplum) [reply Leanplum/Utils/LPExceptionHandler.h]: Nope, we dont care. Unless the copied code is with a external apache license.

File: Leanplum/Leanplum.h (hunk 5 above)
Comment (hrishileanplum) [reply Leanplum/Leanplum.h]: Yes, the goal here is the keep the existing api interfaces as same as possible.

**Pull request merged**